### ISSUE TITLE
Documentation: Full API reference

### ISSUE THREAD

**Issue by zspitz** [xunit/xunit#1902]
There should be a full API reference on the document site, auto-generated in some way.

I imagine each type should have a page, listing all its members; and it should be possible to hyperlink to each member.

The following issues need to be resolved:

- [ ] Should the documentation be generated from one of the platform-specific assembly sets? Or should Roslyn be used to parse the source code?
- [ ] What existing documentation generator can be used for this?
- [ ] What URL structure is appropriate for the API, starting from the docs site?

**Comment by ErikSchierboom** [comment]
I agree that this would be a huge improvement! xUnit is great, but this is one area where I feel NUnit has the edge.

**Comment by zspitz** [comment]
@ErikSchierboom Do you have any experience producing an API reference? Do you have any thoughts about how this should be done?

**Comment by bradwilson** [comment]
Ideally I'd like:

- Auto-generation of the HTML from the XML doc comments
- Ability to have access to multiple versions

Personally I'm okay with the process of kicking off the generation/pushing to the site is something I do by hand, since I don't release new versions very often. I would not do it for CI builds, only builds I ship to NuGet.

**Comment by ErikSchierboom** [comment]
I don't have any experience with generating docs, but I did find this: https://dotnet.github.io/docfx/

**Comment by stakx** [comment]
DocFX is relatively easy to set up. As an example, here's the website it generates for the xunit/xunit repo if you don't apply any customizations (styling, additional content): https://stakx.github.io/xunit/.

In order to get that result, you'd do the following:

1. Install DocFX, e.g. using `choco install docfx`.

2. Go to your local xunit working directory and run `docfx init -q`.

3. Go into the generated `docfx_project` folder (which you can just rename to `docs`) and edit the `docfx.json` configuration file found there:

   <details>
   <summary>Click here to expand the contents of the `docfx.json` file.</summary>

   ```json
   {
   "metadata": [
      {
      "src": [
         {
         "cwd": "..",
         "files": [
               "src/**.csproj"
         ],
         "exclude": [
               "test/**.csproj"
         ]
         }
      ],
      "dest": "api",
      "disableGitFeatures": false,
      "disableDefaultFilter": false
      }
   ],
   "build": {
      "content": [
      {
         "files": [
         "api/**.yml",
         "api/index.md"
         ]
      },
      {
         "files": [
         "articles/**.md",
         "articles/**/toc.yml",
         "toc.yml",
         "*.md"
         ]
      }
      ],
      "resource": [
      {
         "files": [
         "images/**"
         ]
      }
      ],
      "overwrite": [
      {
         "files": [
         "apidoc/**.md"
         ],
         "exclude": [
         "obj/**",
         "_site/**"
         ]
      }
      ],
      "dest": "_site",
      "globalMetadataFiles": [],
      "fileMetadataFiles": [],
      "template": [
      "default"
      ],
      "postProcessors": [],
      "markdownEngineName": "markdig",
      "noLangKeyword": false,
      "keepFileLink": false,
      "cleanupCacheHistory": false,
      "disableGitFeatures": false,
      "xrefService": [ "https://xref.docs.microsoft.com/query?uid={uid}" ]
   }
   }
   ```
   </details>

4. Still inside the `docfx_project` directory, run `docfx -s` from a Visual Studio developer command prompt. This will first generate the site under a `_site` subdirectory and then serve it over `localhost:8080`.

5. You can copy the contents of the `_site` directory over to your website, or to your repo's `gh-pages` branch to have it served via GitHub pages.

Sites generated with DocFX are highly customizable, and I like the fact that you can provide additional conceptual documentation, as well as augment or replace auto-generated API documentation quite easily.

One minor downside that I discovered about DocFX when I [automated doc generation for Moq](https://github.com/stakx/moq4/blob/docfx/build/UpdateGitHubPages.ps1) (not online yet) is that it appears to generate slightly different output on subsequent runs, i.e. it's acts somewhat undeterministically. This makes it a little less useful in an automated setting.

I'm happy to help out if help would be welcome.

**Comment by tommck** [comment]
I love how the response in here is "use intellisense" and "we accept pull requests".

Uh... so.. go to the class type `[` and wade through every single attribute in intellisense to see what the options are?

It requires knowledge in order to use this project.  Knowledge you can't get from the website.  And I can't send a pull request if I'm not already familiar with the product.

**Comment by zspitz** [comment]
@tommck What's being discussed here is auto-generating the documentation from XML docs. I don't think anyone is suggesting that the current situation -- "use Intellisense" is ideal.

I think building such documentation would only require familiarity with a documentation-generating system such as DocFX -- no familiarity with xUnit actually needed., Initially the documentation pages could be generated once and (optionally) incorporated into the project with a PR. Once that's done, it should be relatively simple for @bradwilson to incorporate the doc generation step into the website build.

(Of course, if it was that simple, somebody would have done it...)

**Comment by tommck** [comment]
well, "use Intellisense" has been the answer in these issues for over 4 years

**Comment by zspitz** [comment]
@tommck You obviously feel strongly about the issue; would you consider implementing this? xUnit is an open-source project after all.  :)

**Comment by tommck** [comment]
@zspitz 

> It requires knowledge in order to use this project. Knowledge you can't get from the website. **And I can't send a pull request if I'm not already familiar with the product.**

**Comment by stakx** [comment]
@tommck, it isn't really fair to just complain and expect others to do the work for you, for free. This is an open source project, and most (if not all) people involved probably don't see a single $ for their contributions. Please understand that even if people agree with you, they might assign this issue a lower priority than you do, and they are certainly under no obligation to do any work just because of a complaint.

So, to give you some actionable advice, why not either (1) sponsor someone to do the necessary work (i.e. give them an incentive), or (2) get more familiar with the project yourself and submit your own pull requests?

**Comment by zspitz** [comment]
@tommck

You are correct WRT to conceptual documentation, that it requires knowledge of how to use xUnit, but that's not the only type of documentation there is.

For starters, I think there would be much value in having all the different attributes listed in a table on a single page, with the summary; together with links to pages for each individual attribute, where the parameters could be described. All this information would come from the XMLDoc on the different attributes and their parameters. Generating this doesn't require any deep knowledge of how to use xUnit; only the willingness to learn the basics of an document generation system, and pointing it at a copy of the xUnit dlls. You might also consider writing a simple C# program to read the XMLDoc, parse it, and produce the output yourself.

Another useful piece of information for each attribute would be, where can it be used? On a class? On a member? Grouping the page by attribute usage would be very helpful. Again, producing this page doesn't require any special knowledge of xUnit.

Same goes for the different assertion methods. What does `Assert.Equals` do? `Assert.Not`? Having all the assertion methods in one place would be incredibly useful; but generating such a page, or multiple pages, doesn't depend on special project knowledge.

Between attributes and assertions, I think that covers about 90% of the functionality of xUnit.

After that, I think the next step would be to submit a PR with these additional pages. This does involve a little work -- you have to fork the project, figure out how the xUnit website documentation is creating from source, and insert the new pages into that source,. You probably also would want to test it to see that the links work properly. I won't say it's necessarily trivial to do so, but it's still within the realm of possibility, considering this project is open source.

Optionally, if you could figure out how to auto-generate these pages from the current version, and incorporate that into the website build process, that would be great.

Having done that, you could submit a PR; and considering the level of interest in this issue, the project maintainers would place moving this forward as a high priority. 

With this kind of documentation in hand, it becomes much easier (for you or someone else) to build conceptual documentation. _"xUnit has attributes X and Y for doing A, and M and N, for doing B"_ is much easier to write when I have the list of  attributes in front of me; as is a guide in the form of _"If you want to do A, use X or Y. if you want to do B, use M or N"_.

**Comment by tommck** [comment]
@zspitz Yes, I understand that these are the steps involved, I just find it highly unusual for a project with as much critical mass as this one not to have even basic API documentation on their website. I mean, it seems to be the default for Microsoft recommendations, even over their own framework.

I know this is open source, but to expect a new person to come here wanting to know how to use a library to be the one to document it seems a bit ridiculous to me

**Comment by stakx** [comment]
> to expect a new person to come here wanting to know how to use a library to be the one to document it seems a bit ridiculous to me

And it would be. But this goes both ways: you also cannot expect anyone else to do that work for you. You _could_, however, choose to be the one who gets things going. If you are unwilling to invest in some way, for whatever reason, the only other option is to remain patient.

That being said, I agree with you that any project as popular as xUnit.net ideally should have documentation that's as thorough as can be.

**Comment by tommck** [comment]
> remain patient

🤣

Based on searches in these issues, it's already been more than 4 years that people have been asking for it.  
That's just going to make people choose other options, since there's no "here's how you use it" section on the website

**Comment by PhilT** [comment]
I've just spent a fair bit of time searching for the assertion docs. I found this: https://xunit.net/docs/comparisons#assertions but had to search Google to get there. Couldn't see a link in the doc home page. 

I think having a list of assertions would be a good start.

**Comment by PhilT** [comment]
In addition, I think the home page could do with a facelift. 

Some details at the top really should be at the bottom. The focus should be on what it does, where it can be used, install, usage, and docs.

**Comment by domestomas** [comment]
> I've just spent a fair bit of time searching for the assertion docs. I found this: https://xunit.net/docs/comparisons#assertions but 
That's what i have been looking for. Could anyone add it to the title page please?

**Comment by davidcian** [comment]
Since the response to requests for xUnit documentation is usually "make it yourself" I'll bite and would love to write some documentation myself, but where? Open a pull request, sure, but with what?

**Comment by bradwilson** [comment]
@davidcian If your question is "where does the source for the documentation site", it exists in the `gh-pages` branch of this project: https://github.com/xunit/xunit/tree/gh-pages

If the question is "how do you work on GitHub pages sites locally", then this is the page describing the requirements: https://docs.github.com/en/pages/setting-up-a-github-pages-site-with-jekyll/testing-your-github-pages-site-locally-with-jekyll

If the question is "what tool should I use to convert the XML output into documentation pages", then I don't have an answer for that, as it's part of what needs to be figured out. There is a comment above from @ErikSchierboom about DocFX, which is something that could be investigated: https://github.com/xunit/xunit/issues/1902#issuecomment-466150389

**Comment by AlanMcBee** [comment]
Yesterday, I committed to making my OSS contribution all about the documentation for xUnit. I'm tackling this issue first. Hope to share cool stuff Real Soon Now.

**Comment by bradwilson** [comment]
Closing this as the new site exists at https://api.xunit.net with source at https://github.com/xunit/api.xunit.net

**Comment by ErikSchierboom** [comment]
That is very helpful! Thanks
